internal/nix/command.go

```diff
@@ -1,20 +1,258 @@
+}
+
+func (c *cmd) error(ctx context.Context, err error) error {
+	if err == nil {
+		return nil
+	}
+
+	cmdErr := &cmdError{err: err}
+	if errors.Is(err, exec.ErrNotFound) {
+		cmdErr.msg = fmt.Sprintf("nix: %s not found in $PATH", c.Args[0])
+	}
+
+	switch {
+	case errors.Is(ctx.Err(), context.Canceled):
+		cmdErr.msg = "nix: command canceled"
+	case errors.Is(ctx.Err(), context.DeadlineExceeded):
+		cmdErr.msg = "nix: command timed out"
+	default:
+		cmdErr.msg = "nix: command error"
+	}
+	cmdErr.msg += ": " + c.String()
+
+	var exitErr *exec.ExitError
+	if errors.As(err, &exitErr) {
+		if stderr := c.stderrExcerpt(exitErr.Stderr); len(stderr) != 0 {
+			cmdErr.msg += ": " + stderr
+		}
+		if exitErr.Exited() {
+			cmdErr.msg += fmt.Sprintf(": exit code %d", exitErr.ExitCode())
+			return cmdErr
+		}
+		if stat, ok := exitErr.Sys().(syscall.WaitStatus); ok && stat.Signaled() {
+			cmdErr.msg += fmt.Sprintf(": exit due to signal %d (%[1]s)", stat.Signal())
+			return cmdErr
+		}
+	}
+
+	if !errors.Is(err, ctx.Err()) {
+		cmdErr.msg += ": " + err.Error()
+	}
+	return cmdErr
+}
+
+func (*cmd) stderrExcerpt(stderr []byte) string {
+	stderr = bytes.TrimSpace(stderr)
+	if len(stderr) == 0 {
+		return ""
+	}
+
+	lines := bytes.Split(stderr, []byte("\n"))
+	slices.Reverse(lines)
+	for _, line := range lines {
+		line = bytes.TrimSpace(line)
+		after, found := bytes.CutPrefix(line, []byte("error: "))
+		if !found {
+			continue
+		}
+		after = bytes.TrimSpace(after)
+		if len(after) == 0 {
+			continue
+		}
+		stderr = after
+		break
+
+	}
+
+	excerpt := string(stderr)
+	if !strconv.CanBackquote(excerpt) {
+		quoted := strconv.Quote(excerpt)
+		excerpt = quoted[1 : len(quoted)-1]
+	}
+	return excerpt
+}
+
+type cmdArgs []any
+
+func appendArgs[E any](args cmdArgs, new []E) cmdArgs {
+	for _, elem := range new {
+		args = append(args, elem)
+	}
+	return args
+}
+
+func (c cmdArgs) StringSlice() []string {
+	s := make([]string, len(c))
+	for i := range c {
+		s[i] = fmt.Sprint(c[i])
+	}
+	return s
+}
+
+func (c cmdArgs) String() string {
+	if len(c) == 0 {
+		return ""
+	}
+
+	sb := &strings.Builder{}
+	c.writeQuoted(sb, fmt.Sprint(c[0]))
+	if len(c) == 1 {
+		return sb.String()
+	}
+
+	for _, arg := range c[1:] {
+		sb.WriteByte(' ')
+		c.writeQuoted(sb, fmt.Sprint(arg))
+	}
+	return sb.String()
+}
+
+func (cmdArgs) writeQuoted(dst *strings.Builder, str string) {
+	needsQuote := strings.ContainsAny(str, ";\"'()$|&><` \t\r\n\\#{~*?[=")
+	if !needsQuote {
+		dst.WriteString(str)
+		return
+	}
+
+	canSingleQuote := !strings.Contains(str, "'")
+	if canSingleQuote {
+		dst.WriteByte('\'')
+		dst.WriteString(str)
+		dst.WriteByte('\'')
+		return
+	}
+
+	dst.WriteByte('"')
+	for _, r := range str {
+		switch r {
+		// Special characters inside double quotes:
+		// https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html#tag_02_02_03
+		case '$', '`', '"', '\\':
+			dst.WriteRune('\\')
+		}
+		dst.WriteRune(r)
+	}
+	dst.WriteByte('"')
+}
```

Comment:
sigh, I wish there was some external utility we could rely on for this...
I tried looking but couldn't find a good one.

Reply:
I usually refer to [mvdan.cc/sh/v3] for shell syntax stuff. I looked at the logic in <https://github.com/mvdan/sh/blob/v3.8.0/syntax/quote.go#L48> and just extracted the parts we needed for errors/logging instead of importing the package.

[mvdan.cc/sh/v3]: https://pkg.go.dev/mvdan.cc/sh/v3